semantics/src/traversal/functions.rs

```diff
@@ -57,8 +57,6 @@ pub fn func_def(
             .transpose()?
             .unwrap_or_else(|| Tuple::empty().to_fixed_size());
 
-        function_scope.borrow_mut().return_type = return_type.clone();
```

Comment:
I came across this change today (I missed this in my review...I'm a terrible reviewer btw). I'm not a big fan of this. In fact, when I was rebasing #135 on top of latest master today I accidentally dropped this line and I was baffled for a while why things why functions are reporting the wrong return type. Also having `return_type` on the `BlockScope` doesn't feel right to me as it is an attribute that doesn't mean anything for block scopes that originate in loops or other blocks. I do realize that it could become meaningful some day but currently it isn't.

So, what I'm proposing in this PR instead is to enrich the `BlockScopeType::Function`  to carry the name of the function ``BlockScopeType::Function { name: String }`. That way, we can lookup the defintion of the function in the contract definitions and get rid of `return_type` in `BlockScope`.

The PR has some ugly spots that would need to be cleaned up but I wanted to get a draft of this out to get your opinion.

Reply:
"refactor scopes" is something that's been bouncing around in my head alot lately, so good to see this.

I agree that having `return_type` on the block scope is a bit out of place. In the PR where this was merged, I tried doing something similar to what you described, but ran into issues with the block scope inheritance functions, since they rely on concrete enum variants, so this refactor might be a bit more involved.

Reply:
> but ran into issues with the block scope inheritance functions

Yes, I worked around that by introducing a `inherits_function` method which pattern matches on `BlockScopeType::Function {..}`. That said, it's actually not that useful since every block scope inherits from a function which makes it so that we only actually use that for tests.

>"refactor scopes" is something that's been bouncing around in my head a lot lately

Good! Maybe we stick heads together in the next meeting. I've also thought about more radical changes including traits etc but I didn't arrive on anything I was happy with. Huge part of that is probably due to just missing some good rust design patterns.